### ISSUE TITLE
New IDL interface to Helioviewer API

### ISSUE THREAD

**Issue by dzarro** [Helioviewer-Project/api#2]
I've written a few simple IDL routines that interact with the Helioviewer API documented at http://api.helioviewer.org/docs/v2/

The first is *hv_source.pro* which returns the Source ID for a specified combination of observatory/instrument/detector/measurement as in the following examples:

```
IDL> help,hv_source(obs='SOHO',inst='EIT',meas='284')
<Expression>    LONG64    =                      2
IDL> help,hv_source(obs='SDO',inst='AIA',meas='171')
<Expression>    LONG64    =                     10
IDL> help,hv_source(obs='STEREO_A',inst='secchi',det='euvi',meas='171')
<Expression>    LONG64    =                     20
IDL> help,hv_source(obs='STEREO_B',inst='secchi',det='euvi',meas='195')
<Expression>    LONG64    =                     25
```
 
The second ls *hv_search.pro* which returns metadata (and optionally the header) for the closest JPEG2000 file to a specified time/date and source ID as in the following example:

```
IDL> a=hv_search('01:00 1-may-10',25,header=header)
IDL> help,/st,a
** Structure <1653ab8>, 9 tags, length=88, data length=84, refs=1:
   ID              STRING    '34302878'
   DATE            STRING    '2010-05-01 01:00:53'
   SCALE           DOUBLE           1.5962623
   WIDTH           LONG64                      2048
   HEIGHT          LONG64                      2048
   REFPIXELX       DOUBLE           1035.4200
   REFPIXELY       DOUBLE           1049.6200
   SUNCENTEROFFSETPARAMS
                   OBJREF    <ObjHeapVar5170(LIST)>
   LAYERINGORDER   LONG64                         2
```

The third  is *hv_get.pro* which downloads the closest JPEG2000 file for a specified time/date and source ID as in the following example:

```
IDL> hv_get,'01:00 1-may-10',25,local=local
```

The name of the downloaded file is returned in the variable local:

```
IDL> print,local
/Volumes/moat_disk_1/Users/zarro/data/2010_05_01__01_00_53_637__STEREO-B_SECCHI_EUVI_195.jp2
```

**Comment by wafels** [comment]
We don't have an IDL / Helioviewer client, so this is great!  SunPy has a Helioviewer client, and all that code is kept in the SunPy repository (https://github.com/sunpy/sunpy/blob/master/sunpy/net/helioviewer.py).  I wonder if the SSWIDL / Helioviewer client would be better kept in SSW?

**Comment by dzarro** [comment]
Another useful routine ls *hv_search.pro* which returns metadata for the closest JPEG2000 file to a specified time/date and source ID as in the following example:

```
IDL> a=hv_search('01:00 1-may-10',25)
IDL> help,/st,a
** Structure <1653ab8>, 9 tags, length=88, data length=84, refs=1:
   ID              STRING    '34302878'
   DATE            STRING    '2010-05-01 01:00:53'
   SCALE           DOUBLE           1.5962623
   WIDTH           LONG64                      2048
   HEIGHT          LONG64                      2048
   REFPIXELX       DOUBLE           1035.4200
   REFPIXELY       DOUBLE           1049.6200
   SUNCENTEROFFSETPARAMS
                   OBJREF    <ObjHeapVar5170(LIST)>
   LAYERINGORDER   LONG64                         2
```

Lastly, there is *hv_get.pro* which downloads the closest JPEG2000 file for a specified time/date and source ID as in the following example:

```
IDL> hv_get,'01:00 1-may-10',25,local=local
```

The name of the downloaded file is returned in the variable *local*:

```
IDL> print,local
/Volumes/moat_disk_1/Users/zarro/data/2010_05_01__01_00_53_637__STEREO-B_SECCHI_EUVI_195.jp2
```

**Comment by dzarro** [comment]
The three client programs I developed are currently stored in 
$SSW/gen/idl/clients/hv.  I propose we promote the current Helioviewer 
IDL code in $SSW/packages/helioviewer to the GEN clients area where it 
will be more generally accessible.

**Comment by wafels** [comment]
The code in $SSW/packages/helioviewer seems to be a copy of the JP2Gen code.  It is pretty specialized, and generally speaking, only useful to people wanting to write their own Helioviewer-compliant JPEG2000 files.  Some questions:
(1) What is the purpose of $SSW/gen/idl/clients/hv?
(2) What would be the benefit of the proposed move?
(3) Would the proposed move mean that everyone get the Helioviewer code?

**Comment by dzarro** [comment]
(1) $SSW/gen/idl/clients/hv is a temporary placeholder so that the above client routines will be distributed as part of the SSW/gen tree.
(2) No real benefit of the proposed move since (as you say) $SSW/packages/helioviewer is highly specialized.
(3) If moved to GEN then, the Helioviewer package would automatically be distributed to SSW users. Currently, users have to specifically select the package when updating their SSW installation.

**Comment by dzarro** [comment]
I noticed that some file JPEG2000 files on the server have zero byte size. For example,the following query:

http://api.helioviewer.org/v2/getJP2Image/?date=2015-04-01T00:00:00.000Z&sourceId=10

downloads the file:

2015_04_01__00_00_11_34__SDO_AIA_AIA_171.jp2   0KB

Same problem occurs in VI.  This is causing the IDL client to fail.

**Comment by wafels** [comment]
The sunpy helioviewer client finds the proper non-zero sized file - it does exist on the server:

from sunpy.net import helioviewer
h = helioviewer.HelioviewerClient()
h.download_jp2('2015-04-01T00:00:00.000Z', sourceId=10)
In [4]: h.download_jp2('2015-04-01T00:00:00.000Z', sourceId=10)
Out[4]: u'/Users/ireland/sunpy/data/2015_04_01__00_00_11_34__SDO_AIA_AIA_171.jp2'

The SunPy client uses a version of the API before we made a distinction between v1 and v2. The problem might be there somewhere.

**Comment by dzarro** [comment]
What does:

  ls -al 
/Users/ireland/sunpy/data/2015_04_01__00_00_11_34__SDO_AIA_AIA_171.jp2

return? When I enter the following query in Firefox:

http://api.helioviewer.org/v2/getJP2Image/?date=2015-04-01T00:00:00.000Z&sourceId=10

it downloads a zero byte file.

**Comment by dzarro** [comment]
I wrote another user IDL client routine that returns the JPEG2000 start and end times for a specified source ID:

```
IDL> hv_times,10,tstart,tend,nickname=nickname
IDL> help,tstart,tend,nickname
TSTART          STRING    = '2010-06-02 00:05:36'
TEND            STRING    = '2015-04-08 12:42:35'
NICKNAME        STRING    = 'AIA 171'
```

**Comment by dzarro** [comment]
I have a suggestion for Get Closest Image. Could it include the spacecraft, instrument, detector, and measurement fields in the JSON reponse output?

**Comment by wafels** [comment]
The file exists.  Go to

http://helioviewer.org/?date=2015-04-01T00:00:00.000Z&imageScale=4.8408818&centerX=0.00014773198852539063&centerY=2.4204409&imageLayers=%5BSDO,AIA,AIA,171,1,100%5D&eventLayers=%5BAR,all,1%5D,%5BCC,all,1%5D,%5BCD,all,1%5D,%5BCH,all,1%5D,%5BCJ,all,1%5D,%5BCE,all,1%5D,%5BCR,all,1%5D,%5BCW,all,1%5D,%5BEF,all,1%5D,%5BER,all,1%5D,%5BFI,all,1%5D,%5BFA,all,1%5D,%5BFE,all,1%5D,%5BFL,all,1%5D,%5BLP,all,1%5D,%5BOS,all,1%5D,%5BPG,all,1%5D,%5BSG,all,1%5D,%5BSP,all,1%5D,%5BSS,all,1%5D&eventLabels=true

open up the "Images" accordion and click on the "landscape" icon.  You can download the file directly.  Note that almost all out JP2 files are in greyscale - color tables are applied on the fly.  I also checked that the Helioviewer/SunPy client downloaded the correct 1.1 MB file and that the downloaded image looks like the one available via helioviewer.org.

I have an idea as to what might be going on, but I will need to be at work to try it out.

Please file your suggestion for Get Closest Image in a separate issue.  This will help keep the API suggestion and the bug separate.

**Comment by wafels** [comment]
Closed - see #3
